mmv1/products/storage/ManagedFolder.yaml

```diff
@@ -0,0 +1,104 @@
+# Copyright 2024 Google Inc.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+--- !ruby/object:Api::Resource
+name: 'ManagedFolder'
+kind: 'storage#managedFolder'
+base_url: 'b/{{bucket}}/managedFolders'
+self_link: 'b/{{bucket}}/managedFolders/{{%name}}'
+id_format: '{{bucket}}/{{name}}'
+has_self_link: true
+immutable: true
+skip_sweeper: true # Skipping sweeper since this is a child resource.
+description: |
+  A Google Cloud Storage Managed Folder.
+
+  You can apply Identity and Access Management (IAM) policies to
+  managed folders to grant principals access only to the objects
+  within the managed folder, which lets you more finely control access
+  for specific data sets and tables within a bucket. You can nest
+  managed folders up to 15 levels deep, including the parent managed
+  folder.
+
+  Managed folders can only be created in buckets that have uniform
+  bucket-level access enabled.
+references: !ruby/object:Api::Resource::ReferenceLinks
+  guides:
+    'Official Documentation': 'https://cloud.google.com/storage/docs/managed-folders'
+  api: 'https://cloud.google.com/storage/docs/json_api/v1/managedFolder'
+iam_policy: !ruby/object:Api::Resource::IamPolicy
+  allowed_iam_role: 'roles/storage.objectViewer'
+  admin_iam_role: 'roles/storage.admin'
+  parent_resource_attribute: 'managed_folder'
+  base_url: 'b/{{bucket}}/managedFolders/{{%managed_folder}}'
```

Comment:
Hi @melinath, @trodge. Just wanted to confirm that Here path escape character(%) works or not? 
I looked at some generated IAM files and found that url is not being constructed with url encoding of path parameters. For example, [this code](https://github.com/hashicorp/terraform-provider-google/blob/main/google/services/sourcerepo/iam_sourcerepo_repository.go#L212), I can see that `urlTemplate` is being constructed directly from the parameters value without any url encoding. 
I do see all the generated tests passed but It might be because of "managed-folder-1/" resource name. We basically don't need any encoding here as there is only one slash here which is at the end. It may fail for nested managed folders like "a/b/".

Reply:
+1 that we should test the IAM behavior with nested folders to be safe. The code [looks correct to me](https://github.com/hashicorp/terraform-provider-google/blob/a4438d4339a4386b018c63d01cd43a97dc4ab431/google/services/sourcerepo/iam_sourcerepo_repository.go#L212) but also I don't see any difference from the google_sourcerepo_repository_iam code you linked (and that [doesn't override the base_url for IAM at all](https://github.com/GoogleCloudPlatform/magic-modules/blob/c07e40b24de7aeed6faa445a659ea5ff28ff50f6/mmv1/products/sourcerepo/Repository.yaml#L29).) It could be that the % is a no-op here.

Reply:
Actually I took this url constructor reference just to ask that I don't see any url encoding happening in the IAM resource. As you show that MMV1 yaml is not having any path escaper for IAM, it won't have any logic in generated resource but I tried manually with `%` character and No difference. It is generating urlTemplate like `{{StorageBasePath}}b/a/managedFolders/b//iam` which I think doesn't make any sense and results in 404 not found error.

Reply:
I just tested this locally and you're absolutely right.  The current approach doesn't work for nested folders. I don't see any hooks in [iam_policy.erb](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/templates/terraform/iam_policy.go.erb#L307) to modify the URL, so it looks like we'll have to write it manually.

Is there any documentation on how create a handwritten IAM policy resource? Should I start from the generated code and update it manually?

Reply:
So if I understand correctly, the problem is that the managed folder path has to end with a "/" (ideally, if we could set up validation for the parent resource field), but when you join that with the IAM methods it ends up having an extra slash? If that's correct, you could try adding setting the method separator to an empty string, since the trailing "/" would act as the separator on its own.

Reply:
oh - no, the problem is that the managed folder path is supposed to be URL escaped.

Reply:
Unfortunately, as far as I know, the `%` just indicates that the given part of the path can contain slashes; it doesn't mean that we urlencode that part of the path. If that's necessary here, then either users need to provide the path already escaped (for example using the Terraform [`urlencode`](https://developer.hashicorp.com/terraform/language/functions/urlencode) function might work?) or the IAM support for this resource needs to be handwritten.

Adding support for this as a feature of the generator would be a pretty significant lift, I think, so it is not something I'd want to commit to in the context of this PR.

Reply:
Asking the user to urlencode might break other things. Let me give it a try and I'll get back to you

Reply:
+1. I think, we should do url encoding from provider side. As it appears This is the special case for IAM auto generation, We can solve this by moving this resource to Handwritten and applying custom encoding logic.

Reply:
@melinath, Can it be fixed with MMV1 auto generation? 
We may require this url encoding(path escape) with upcoming new features. If it is possible to add path escape support in auto generated IAM resource then we won't require moving IAM resource to handwritten and no need to write custom logic.

Reply:
If It requires significant effort to modify generator then we should move IAM to handwritten by following guidelines given [here](https://googlecloudplatform.github.io/magic-modules/develop/resource/#add-iam-support). 
In HandWritten resource, I think we need to only change `qualifyManagedFolderUrl` function with path escaping logic. Probably Something like this should work,
`urlTemplate := fmt.Sprintf("{{StorageBasePath}}%s/%s", fmt.Sprintf("b/%s/managedFolders/%s", u.bucket, url.PathEscape(u.managedFolder)), methodIdentifier)`

Reply:
Yeah, we probably don't want to handle this specially for IAM right now. If there are additional resources that need it we should discuss that separately (so this resources isn't blocked) and you can look into converting back to generated later.